Document CategoricalString

alyst opened the pull request:
Updated the docs to reflect the changes brought by #87.

As with the introduction of `CategoricalString` some parts of docs are becoming cumbersome, I've restructured the text a little bit, so that the flow is smooth.

I've tried to make these changes atomic, so if there are concerns regarding some of the changes and their relevance to this PR, it could be addressed on per-commit basis.
Then the commits could be merged into mainline without squashing.

nalimilan commented on docs/src/using.md:
```diff
@@ -10,9 +10,9 @@ julia> using CategoricalArrays
 4-element CategoricalArrays.CategoricalArray{String,1,UInt32}:
  "Old"   
- "Young" 
+ "Young"
```
Are you sure it's a good idea to remove these? As you can see, I just copy/pasted the REPL output, so we're certain the doctests will pass. I don't see the point of removing them.

alyst replied:
oops. It's not me it's Atom :) I didn't realize it's padding, will revert

nalimilan commented on docs/src/index.md:
```diff
@@ -2,10 +2,8 @@
 This package provides a replacement for [DataArrays.jl](https://github.com/JuliaStats/DataArrays.jl)'s `PooledDataArray` type. Contrary to that type, it supports both non-nullable and nullable arrays (i.e. arrays allowing for missing data), using the [`Null`](https://github.com/JuliaData/Nulls.jl) type. It is also based on a simpler design by only supporting categorical data, which allows offerin ...
 
-The package provides the `CategoricalArray` type designed to hold categorical data (either unordered/nominal or ordered/ordinal) efficiently and conveniently. `CategoricalArray{T}` holds values of type `T`. `CategoricalArray{Union{Null, T}}`, either holds values of type `T`, or missing values (represented as `null`, of the [`Null`](https://github.com/JuliaData/Nulls.jl) type). When indexed, `Cate ...
+The package provides the `CategoricalArray` type designed to hold categorical data (either unordered/nominal or ordered/ordinal) efficiently and conveniently. `CategoricalArray{T}` holds values of type `T`. `CategoricalArray{Union{Null, T}}` variant can also contain missing values (represented as `null`, of the [`Null`](https://github.com/JuliaData/Nulls.jl) type). When indexed, `CategoricalArray ...
```
Add "The" at the beginning of sentences where appropriate. "Allows" was better than "provides" IMHO, but maybe we can find another term.

nalimilan commented on docs/src/index.md:
```diff
@@ -2,10 +2,8 @@
+The package provides the `CategoricalArray` type designed to hold categorical data (either unordered/nominal or ordered/ordinal) efficiently and conveniently. `CategoricalArray{T}` holds values of type `T`. `CategoricalArray{Union{Null, T}}` variant can also contain missing values (represented as `null`, of the [`Null`](https://github.com/JuliaData/Nulls.jl) type). When indexed, `CategoricalArray ...
 
-`CategoricalValue` objects are simple wrappers around the actual categorical levels which allow for very efficient extraction and equality tests. Indeed, the main feature of categorical arrays types is that they store a pool of the levels which can appear in the variable. These levels are stored in a specific order: for unordered arrays, this order is only used for pretty printing (e.g. in cross  ...
+Indeed, the main feature of `CategoricalArray` is that it maintains a pool of the levels which can appear in the data. These levels are stored in a specific order: for unordered arrays, this order is only used for pretty printing (e.g. in cross tables or plots); for ordered arrays, it also allows comparing values using the `<` and `>` operators: the comparison is then based on the ordering of lev ...
```
"The `levels!` function". "Levels are also automatically expanded".

nalimilan commented on docs/src/implementation.md:
```diff
@@ -1,16 +1,21 @@
+`CategoricalArray` is made of the two fields:
 
-`CategoricalArray` is made of two fields:
+- `refs`: an integer array that stores the indices of levels for each `CategoricalArray` element; `0` denotes a missing value (for `CategoricalArray{Union{T, Null}}` only).
```
"indices of levels" could be misleading since these are not the positions of the elements in the levels, but in the index. I"d rather say something like "that stores the position of the level in the pool's index for each element". Maybe we should also point to the text below which describes the index.

alyst replied:
btw, the corresponding field and method in `CategoricalString/CategoricalValue` are called `level`. To avoid confusion I would rather call them either `ref` (it might be confused with `Base.Ref`, but it's consistent with `reftype()` and `CategoricalArray.refs`) or `index`.

nalimilan commented on docs/src/implementation.md:
```diff
@@ -1,16 +1,21 @@
 
-- `refs`: an integer vector giving the index of the level in the pool for each element; if the array is nullable, `0` indicates a missing value.
-- `pool`: the `CategoricalPool` object keeping the levels of the array.
+`CategoricalPool{V,R,C}` type keeps track of the levels of type `V` and associates them with an integer reference codes of type `R` (for internal use). It offers methods to set the levels, change their order while preserving the references, and efficiently get the integer index corresponding to a level and vice-versa. Whether the values of `CategoricalArray` are ordered or not is stored in an `or ...
```
"The `CategoricalPool{V,R,C}` type". "an integer reference code".

nalimilan commented on docs/src/implementation.md:
```diff
@@ -1,16 +1,21 @@
 
-Whether an array (and its values) are ordered or not is stored as a property of the pool.
+The type parameters of `CategoricalArray{T, N, R <: Integer, V, U}` are a bit complex:
+ - `T` is the actual element type of the array, ignoring the `CategoricalValue` wrapper; if `T >: Null`, then the array is nullable.
```
It feels weird to say "actual" since that type doesn't exist anywhere.

nalimilan commented on docs/src/implementation.md:
```diff
@@ -1,16 +1,21 @@
+The type parameters of `CategoricalArray{T, N, R <: Integer, V, U}` are a bit complex:
+ - `T` is the actual element type of the array, ignoring the `CategoricalValue` wrapper; if `T >: Null`, then the array is nullable.
+ - `N` is the number of array dimensions.
+ - `R` is the reference type, the element type of the `refs` field; it allows to optimize memory usage depending on the number of levels (i.e. `CategoricalArray` with less than 256 levels can use `R = UInt8`).
```
"it allows optimizing"

nalimilan commented on docs/src/implementation.md:
```diff
@@ -1,16 +1,21 @@
+ - `U` can be either `Union{}` for non-nullable arrays, or `Null` for nullable arrays.
 
-The type parameters of `CategoricalArray{T, N, R <: Integer, U, V}` are a bit complex. `T` corresponds to the element type of the array, ignoring the `CategoricalValue` wrapper; if `T >: Null`, then the array is nullable. `N` corresponds to the number of dimensions of the array. `R` is the reference type, i.e. the element type of the `refs` field. `V` corresponds to the type of the levels, which  ...
+Only `T`, `N` and `R` could be specified upon construction. The last 2 parameters are chosen automatically, but are needed for the definition of the type. In particular, `U` allows expressing that `CategoricalArray{T, N}` inherits from `AbstractArray{Union{CategoricalValue{V, R}, U}, N}` (which is equivalent to `AbstractArray{CategoricalValue{V, R}, N}` for non-nullable arrays, and to `AbstractAr ...
```
"last two"

nalimilan commented on docs/src/implementation.md:
```diff
@@ -7,14 +7,15 @@
  - `U` can be either `Union{}` for non-nullable arrays, or `Null` for nullable arrays.
 
-Only `T`, `N` and `R` could be specified upon construction. The last 2 parameters are chosen automatically, but are needed for the definition of the type. In particular, `U` allows expressing that `CategoricalArray{T, N}` inherits from `AbstractArray{Union{CategoricalValue{V, R}, U}, N}` (which is equivalent to `AbstractArray{CategoricalValue{V, R}, N}` for non-nullable arrays, and to `AbstractAr ...
+Only `T`, `N` and `R` could be specified upon construction. The last 3 parameters are chosen automatically, but are needed for the definition of the type. In particular, `U` allows expressing that `CategoricalArray{T, N}` inherits from `AbstractArray{Union{C, U}, N}` (which is equivalent to `AbstractArray{C, N}` for non-nullable arrays, and to `AbstractArray{Union{C, Null}, N}` for nullable array ...
```
"three"

nalimilan commented on docs/src/using.md:
```diff
@@ -124,7 +124,7 @@ levels!
 ## Handling Missing Values
 
-The examples above assumed that the data contained no missing values. This is generally not the case in real data. This is where `CategoricalArray{Union{T, Null}}` comes into play. It is essentially the categorical-data equivalent of `Array{Union{T, Null}}`. It behaves exactly the same as `CategoricalArray{T}`, except that when indexed it returns either a `CategoricalValue{T}`, or `null` if a val ...
+The examples above assumed that the data contained no missing values. This is generally not the case in real data. This is where `CategoricalArray{Union{T, Null}}` comes into play. It is essentially the categorical-data equivalent of `Array{Union{T, Null}}`. It behaves exactly as `CategoricalArray{T}`, except that when indexed it returns either a categorical value object (`CategoricalValue{T}` or ...
```
I would mention `CategoricalString` first.

alyst commented:
I've addressed the comments + a few minor tweaks. These commits are really atomic, I'll squash them into appropriate ones after the review.
By "squashing a bit more" you mean squashing all `CategoricalString`-related into a single commit?

nalimilan commented on docs/src/index.md:
```diff
@@ -2,10 +2,8 @@
+The package provides the `CategoricalArray` type designed to hold categorical data (either unordered/nominal or ordered/ordinal) efficiently and conveniently. `CategoricalArray{T}` holds values of type `T`. The `CategoricalArray{Union{Null, T}}` variant can also contain missing values (represented as `null`, of the [`Null`](https://github.com/JuliaData/Nulls.jl) type). When indexed, `CategoricalA ...
 
-`CategoricalValue` objects are simple wrappers around the actual categorical levels which allow for very efficient extraction and equality tests. Indeed, the main feature of categorical arrays types is that they store a pool of the levels which can appear in the variable. These levels are stored in a specific order: for unordered arrays, this order is only used for pretty printing (e.g. in cross  ...
+Indeed, the main feature of `CategoricalArray` is that it maintains a pool of the levels which can appear in the data. These levels are stored in a specific order: for unordered arrays, this order is only used for pretty printing (e.g. in cross tables or plots); for ordered arrays, it also allows comparing values using the `<` and `>` operators: the comparison is then based on the ordering of lev ...
```
My comment still holds here: "the `levels!` function"; "Levels are also automatically expanded".

alyst replied:
I'm torn between "expanded" and "extended". [These tips](https://www.dailywritingtips.com/expanded-and-extended/) do not help the decision much. :)
Will use "expanded".

nalimilan replied:
Hmm, "extended" sounds more correct. Indeed we change `length(levels(x))`, not `area(levels(x))`. :-)

nalimilan commented on docs/src/implementation.md:
```diff
@@ -1,16 +1,22 @@
 
-Whether an array (and its values) are ordered or not is stored as a property of the pool.
+The type parameters of `CategoricalArray{T, N, R <: Integer, V, C, U}` are a bit complex:
+ - `T` is the effective type of array elements unwrapped from `CategoricalString`/`CategoricalValue` helpers; if `T >: Null`, then the array is nullable.
```
"effective" isn't really different from "actual". Better not use any adjective. BTW, "helpers"  would better be "wrappers", though it's redundant with "unwrapped", which could be changed to "without" or something like that.

alyst replied:
From google translate, one of the meanings: *"Effective: fulfilling a specified function in fact, though not formally acknowledged as such"* I think it's fits pretty well, but I can remove if you disagree.

I can either just delete "helpers" or "stripped of CatString/CatValue wrappers" or "without CatString/CatValue wrappers". Which one do you like? :)

nalimilan replied:
That definition doesn't really apply AFAIK, since indexing into the array will return categorical values, so the retrieved elements do not act as `T` elements for that "specified function".

Regarding the second point, "without" sounds simpler to me.

alyst replied:
The *"specified function"* here IMO is the efficient storage of `T` elements, while *formally* indexing returns categorical values. But I'm not bound to that adjective, will remove; "without CatString/CatValue wrappers" is already clear enough.

alyst commented:
@nalimilan Is it good to go?

nalimilan commented:
Thanks!
